Feat/collaboration service

zechajw opened the pull request:
- [X] add routes to check if a user is already in a room
- [X] add routes for a user to join or leave a room
- [x] add socket event to broadcast join/leave to other users in room
- [X] add integration to questions service to fetch question options
- [X] add ability to join room and leave room as shown in the InfoBar

Known Issues:
- [x] Bug with fetching users where user does not appear until page is reloaded 
- [x] Inability to interface with socket api for live data fetching in questions and users
- [x] Poor handling of error and loading states in collaboration 
- [x] Option Dropdown is transparent and interacts weirdly with question details

zechajw committed:
feat: add route to check if user in room and add support fo qn in backend

zechajw committed:
fix: remove console.log

zechajw committed:
feat: add logic to disconnect user from room when leaving

joelwongjy committed:
feat: integrate question filters api to fetch categories

joelwongjy committed:
refactor: improve select preference refresh logic

joelwongjy committed:
fix: get rid of white bars when select expands

zechajw committed:
feat: integrate questions and users into room

zechajw committed:
Merge branch 'staging' into feat/collaboration-service

zechajw committed:
fix: fix bug where menu was transparent

zechajw committed:
feat: add live updates when question is changed

Josephaedan committed:
Commit

zechajw committed:
feat: fix bugs and add support for language changing

zechajw committed:
fix: remove console.log

zechajw committed:
fix: fix bug with userinfocard

zechajw committed:
Change question details component to use react-query

raynerljm commented on backend/collaboration-service/app/src/helpers/socket.helper.ts:
```diff
@@ -52,11 +53,28 @@ export function handlePushUpdates(socket: Socket, version: number, docUpdatesDat
+}
+
+export function handleLanguageChange(socket: Socket, roomId: string, language: LanguageKeyType) {
+    console.log("IM HERE")
```
nit: unused console log

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on backend/collaboration-service/app/src/server.ts:
```diff
@@ -17,20 +22,31 @@ const io = new Server(server, {
   cors: {
     credentials: true,
-    origin: process.env.FRONTEND_ORIGIN,
+    // origin: "*",
```
nit: remove unused code

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

zechajw committed:
feat: add icon to indicate online or offline

raynerljm commented on frontend/app/src/pages/room/[roomId].tsx:
```diff
@@ -1,13 +1,24 @@
   const { roomId } = useParams();
 
-  // TODO: Add Error message if room id not passed as prop
+  const { isLoading, isError, data } = useGetRoomInfo(roomId ?? "")
```
nit: personally i would
1. not pass `?? ""` and just instead pass it as `roomId`
2. then in the `useGetRoomInfo` hook, it would receive `roomId: string | undefined`
3. so the `enabled: roomId !== undefined` means that the fetch will only happen once the page retrieves the path params and passes it into the hook

it's the same behaviour as what you did but the way you handled it might not be as clear. up to you whether you wna change, just commenting

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/api/useGetUserInfo.ts:
```diff
@@ -0,0 +1,34 @@
+
+const getUserInfoResponseSchema = makeSuccessResponseSchema(
+    z.object({
+        user: z.object({
```
is there a reason you didn't use the `userSchema` as defined in `types/user.ts`?

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -1,29 +1,38 @@
   const [socket, setSocket] = useState<Socket | null>(null);
-  const [currentLanguage, setCurrentLanguage] = useState(DEFAULT_LANGUAGE);
+  const [currentLanguage, setCurrentLanguage] = useState(LANGUAGES[language] ?? DEFAULT_LANGUAGE);
+  const [activeQuestionId, setActiveQuestionId] = useState(questionId ?? 0);
```
is there a reason you used `0` as the default value?

zechajw replied:
Fixed in to remove 0 [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -33,71 +42,80 @@ export const Collaborator = ({ roomId }: CollaboratorProps) => {
         size="sm"
         title="Language"
         placeholder="Select language"
+        value={currentLanguage ?? DEFAULT_LANGUAGE}
```
if your `currentLanguage` state already has a default there shouldn't be a `?? DEFAULT_LANGUAGE` here

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -33,71 +42,80 @@ export const Collaborator = ({ roomId }: CollaboratorProps) => {
-        placeholder="Select question"
-        options={[]}
+        placeholder="Select Question"
+        value={activeQuestionId ?? undefined}
```
i don't think you need `?? undefined` here

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/code/QuestionDetails.tsx:
```diff
@@ -1,13 +1,35 @@
+    if (isError || !data) {
+        return (
+            <Box flex={1} padding={"0.75rem"} border="1px" borderColor={"ActiveBorder"} rounded="md" height="full">
+                <Text whiteSpace={"pre-wrap"}>{"Choose a Question"}</Text>
```
nit: don't need `pre-wrap` here

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/code/QuestionDetails.tsx:
```diff
@@ -1,13 +1,35 @@
+    if (isLoading) {
+        return (
+            <Box flex={1} padding={"0.75rem"} border="1px" borderColor={"ActiveBorder"} rounded="md" height="full">
+                <Text whiteSpace={"pre-wrap"}>{"Loading..."}</Text>
```
nit: don't need `pre-wrap` here

zechajw replied:
Fixed in [d5bf5bf](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/d5bf5bf3e84c127ee09ca3aafed19e1d9e9fe361)

raynerljm commented on frontend/app/src/features/room/components/info/RoomUserInfo.tsx:
```diff
@@ -0,0 +1,32 @@
+            <Avatar name={name} src={avatarUrl} />
+            <VStack alignItems="start" gap="0">
+                <Text textStyle="heading-xs">{name ?? email}</Text>
+                {connected ? <Text>Online</Text> : <Text>Offline</Text>}
```
might be nice to have grayed out text for offline/ green text for online

zechajw replied:
Initial change made in [8fb0ae6](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g03/pull/34/commits/8fb0ae6cd894901988191d3a56ec92e1601cd445) but up for further changes

raynerljm reviewed (approved):
lgtm! 🎉 just some nits about handling of `undefined` and `null` states in react as discussed

zechajw committed:
fix: fix issues raised in PR review

- remove unused console.logs and code
- better handling of undefined values in useStates
- use predefined user schema in types/user.ts
- remove unneeded pre-wrap in questiondetails

zechajw committed:
fix: fix bug with userschema validation in RoomUserInfo

joelwongjy committed:
fix: update matching service property names
